crates/bevy_ui/src/interaction_states.rs

```diff
@@ -0,0 +1,61 @@
+/// This module contains components that are used to track the interaction state of UI widgets.
+///
+// Note to implementers: This uses a combination of both marker components and newtype components
+// containing a bool. Markers are used for one-way binding, "write-only" components
+// (like `InteractionDisabled`) that relay instructions from the user to the framework, whereas
+// newtype components are used to request state updates from the framework, which mutates the
+// content of those components on update.
+use bevy_a11y::AccessibilityNode;
+use bevy_ecs::{
+    component::{Component, HookContext},
+    world::DeferredWorld,
+};
+
+/// A marker component to indicate that a widget is disabled and should be "grayed out".
+/// This is used to prevent user interaction with the widget. It should not, however, prevent
+/// the widget from being updated or rendered, or from acquiring keyboard focus.
+///
+/// For apps which support a11y: if a widget (such as a slider) contains multiple entities,
+/// the `InteractionDisabled` component should be added to the root entity of the widget - the
+/// same entity that contains the `AccessibilityNode` component. This will ensure that
+/// the a11y tree is updated correctly.
+#[derive(Component, Debug, Clone, Copy)]
+#[component(on_add = on_add_disabled, on_remove = on_remove_disabled)]
+pub struct InteractionDisabled;
+
+// Hook to set the a11y "disabled" state when the widget is disabled.
+fn on_add_disabled(mut world: DeferredWorld, context: HookContext) {
+    let mut entt = world.entity_mut(context.entity);
+    if let Some(mut accessibility) = entt.get_mut::<AccessibilityNode>() {
+        accessibility.set_disabled();
+    }
+}
+
+// Hook to remove the a11y "disabled" state when the widget is enabled.
+fn on_remove_disabled(mut world: DeferredWorld, context: HookContext) {
+    let mut entt = world.entity_mut(context.entity);
+    if let Some(mut accessibility) = entt.get_mut::<AccessibilityNode>() {
+        accessibility.clear_disabled();
+    }
+}
+
+/// Component that indicates whether a button is currently pressed. This will be true while
+/// a drag action is in progress.
+#[derive(Component, Default, Debug)]
+pub struct ButtonPressed(pub bool);
```

Comment:
Given that the other "bool state" components are immutable, maybe we should make this immutable too for consistency / to enable reliable observer-driven management?